README
NAME
    Array::Iterator - A simple class for iterating over Perl arrays

VERSION
    version 0.10

SYNOPSIS
      use Array::Iterator;

      # create an iterator with an array
      my $i = Array::Iterator->new(1 .. 100);

      # create an iterator with an array reference
      my $i = Array::Iterator->new(\@array);

      # create an iterator with a hash reference
      my $i = Array::Iterator->new({ __array__ => \@array });

      # a base iterator example
      while ($i->has_next()) {
          if ($i->peek() < 50) {
              # ... do something because
              # the next element is over 50
          }
          my $current = $i->next();
          # ... do something with current
      }

      # shortcut style
      my @accumulation;
      push @accumulation => { item => $iterator->next() } while $iterator->has_next();

      # C++ ish style iterator
      for (my $i = Array::Iterator->new(@array); $i->has_next(); $i->next()) {
        my $current = $i->current();
        # .. do something with current
      }

      # common perl iterator idiom
      my $current;
      while ($current = $i->get_next()) {
        # ... do something with $current
      }

DESCRIPTION
    This class provides a very simple iterator interface. It is is
    uni-directional and can only be used once. It provides no means of
    reverseing or reseting the iterator. It is not recommended to alter the
    array during iteration, however no attempt is made to enforce this
    (although I will if I can find an efficient means of doing so). This
    class only intends to provide a clear and simple means of generic
    iteration, nothing more (yet).

METHODS
  Public Methods
    new (@array | $array_ref | $hash_ref)
        The constructor can be passed either a plain perl array, an array
        reference, or a hash reference (with the array specified as a single
        key off the hash, __array__). Single element arrays are not
        supported by either of the first two calling conventions, since it
        is not possible to distinguish between an array of a single element
        which happens to be an array reference, and an array reference of a
        single element, thus previous versions of the constructor would
        raise an exception. If you expect to pass arrays to the constructor
        which may have only a single element, then the array can be passed
        as the element of a HASH reference, with the key, __array__:

          my $i = Array::Iterator->new({ __array__ => \@array });

    has_next([$n])
        This methods returns a boolean. True (1) if there are still more
        elements in the iterator, false (0) if there are not.

        Takes an optional positive integer (> 0) that specifies the position
        you want to check. This allows you to check if there an element at
        arbitrary position. Think of it as an ordinal number you want to
        check:

          $i->has_next(2);  # 2nd next element
          $i->has_next(10); # 10th next element

        Note that has_next(1) is the same as "has_next()".

        Throws an exception if $n <= 0.

    next
        This method returns the next item in the iterator, be sure to only
        call this once per iteration as it will advance the index pointer to
        the next item. If this method is called after all elements have been
        exhausted, an exception will be thrown.

    get_next
        This method returns the next item in the iterator, be sure to only
        call this once per iteration as it will advance the index pointer to
        the next item. If this method is called after all elements have been
        exhausted, it will return undef.

        This method was added to allow for a faily common perl iterator
        idiom of:

          my $current;
          while ($current = $i->get_next()) {
              ...
          }

        In this the loop terminates once $current is assigned to a false
        value. The only problem with this idiom for me is that it does not
        allow for undefined or false values in the iterator. Of course, if
        this fits your data, then there is no problem. Otherwise I would
        recommend the "has_next"/"next" idiom instead.

    peek([$n])
        This method can be used to peek ahead at the next item in the
        iterator. It is non-destructuve, meaning it does not advance the
        internal pointer. If this method is called and attempts to reach
        beyond the bounds of the iterator, it will return undef.

        Takes an optional positive integer (> 0) that specifies how far
        ahead you want to peek:

          $i->peek(2);  # gives you 2nd next element
          $i->peek(10); # gives you 10th next element

        Note that peek(1) is the same as "peek()".

        Throws an exception if $n <= 0.

        NOTE: Prior to version 0.03 this method would throw an exception if
        called out of bounds. I decided this was not a good practice, as it
        made it difficult to be able to peek ahead effectively. This not the
        case when calling with an argument that is <= 0 though, as it's
        clearly a sign of incorrect usage.

    current
        This method can be used to get the current item in the iterator. It
        is non-destructive, meaning that it does not advance the internal
        pointer. This value will match the last value dispensed by "next" or
        "get_next".

    current_index
        This method can be used to get the current index in the iterator. It
        is non-destructive, meaning that it does not advance the internal
        pointer. This value will match the index of the last value dispensed
        by "next" or "get_next".

    get_length
        This is a basic accessor for getting the length of the array being
        iterated over.

  Protected Methods
    These methods are *protected*, in the Java/C++ sense of the word. They
    can only be called internally by subclasses of Array::Iterator, an
    exception is thrown if that condition is violated. They are documented
    here only for people interested in subclassing Array::Iterator.

    _current_index
        An lvalue-ed subroutine which allows access to the iterator's
        internal pointer.

    _iteratee
        This returns the item being iteratated over, in our case an array.

    _get_item ($iteratee, $index)
        This method is used by all other routines to access items with.
        Given the iteratee and an index, it will return the item being
        stored in the $iteratee at the index of $index.

TO DO
    Improve BiDirectional Test suite
        I want to test the back and forth a little more, make sure they work
        well with one another.

    Other Iterators
        Array::Iterator::BiDirectional::Circular, Array::Iterator::Skipable
        and Array::Iterator::BiDirectional::Skipable are just a few ideas I
        have had. I am going to hold off for now until I am sure they are
        actually useful.

BUGS
    None that I am aware of. The code is pretty thoroughly tested (see "CODE
    COVERAGE" below) and is based on an (non-publicly released) module which
    I had used in production systems for about 2 years without incident. Of
    course, if you find a bug, let me know, and I will be sure to fix it.

CODE COVERAGE
    I use Devel::Cover to test the code coverage of my tests, below is the
    Devel::Cover report on this module's test suite.

     ------------------------------- ------ ------ ------ ------ ------ ------ ------
     File                              stmt   bran   cond    sub    pod   time  total
     ------------------------------- ------ ------ ------ ------ ------ ------ ------
     Array/Iterator.pm                100.0  100.0   66.7  100.0  100.0   67.6   98.2
     Array/Iterator/BiDirectional.pm  100.0  100.0    n/a  100.0  100.0   20.2  100.0
     Array/Iterator/Circular.pm       100.0  100.0    n/a  100.0  100.0    7.1  100.0
     Array/Iterator/Reusable.pm       100.0    n/a    n/a  100.0  100.0    5.0  100.0
     ------------------------------- ------ ------ ------ ------ ------ ------ ------
     Total                            100.0  100.0   66.7  100.0  100.0  100.0   99.0
     ------------------------------- ------ ------ ------ ------ ------ ------ ------

SEE ALSO
    This module now includes several subclasses of Array::Iterator which add
    certain behaviors to Array::Iterator, they are:

    Array::Iterator::BiDirectional
        Adds the ability to move backwards and forwards through the array.

    Array::Iterator::Circular
        When this iterator reaches the end of its list, it will loop back to
        the start again.

    Array::Iterator::Reusable
        This iterator can be reset to its beginning and used again.

    The Design Patterns book by the Gang of Four, specifically the Iterator
    pattern.

    Some of the interface for this class is based upon the Java Iterator
    interface.

OTHER ITERATOR MODULES
    There are a number of modules on CPAN with the word Iterator in them.
    Most of them are actually iterators included inside other modules, and
    only really useful within that parent modules context. There are however
    some other modules out there that are just for pure iteration. I have
    provided a list below of the ones I have found, if perhaps you don't
    happen to like the way I do it.

    Tie::Array::Iterable
        This module ties the array, something we do not do. But it also
        makes an attempt to account for, and allow the array to be changed
        during iteration. It accomplishes this control because the
        underlying array is tied. As we all know, tie-ing things can be a
        performance issue, but if you need what this module provides, then
        it will likely be an acceptable compromise. Array::Iterator makes no
        attempt to deal with this mid-iteration manipulation problem. In
        fact it is recommened to not alter your array with Array::Iterator,
        and if possible we will enforce this in later versions.

    Data::Iter
        This module allows for simple iteratation over both hashes and
        arrays. It does it by importing several functions which can be used
        to loop over either type (hash or array) in the same way. It is an
        interesting module, it differs from Array::Iterator in paradigm
        (Array::Iterator is more OO) as well as in intent.

    Class::Iterator
        This is essentially a wrapper around a closure based iterator. This
        method can be very flexible, but at times is difficult to manage due
        to the inherent complextity of using closures. I actually was a
        closure-as-iterator fan for a while, but eventually moved away from
        it in favor of the more plain vanilla means of iteration, like that
        found Array::Iterator.

    Class::Iter
        This is part of the Class::Visitor module, and is a Visitor and
        Iterator extensions to Class::Template. Array::Iterator is a
        standalone module not associated with others.

    Data::Iterator::EasyObj
        Data::Iterator::EasyObj makes your array of arrays into iterator
        objects. It also has the ability to further nest additional data
        structures including Data::Iterator::EasyObj objects.
        Array::Iterator is one dimensional only, and does not attempt to do
        many of the more advanced features of this module.

ACKNOWLEDGEMENTS
    Thanks to Hugo Cornelis for pointing out a bug in "peek()"
    Thanks to Phillip Moore for providing the patch to allow single element
    iteration through the hash-ref constructor parameter.

ORIGINAL AUTHOR
    stevan little, <stevan@iinteractive.com>

ORIGINAL COPYRIGHT AND LICENSE
    Copyright 2004, 2005 by Infinity Interactive, Inc.

    <http://www.iinteractive.com>

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Array-Iterator

0.10     2013-09-18 (SHARYANTO)

         - No functional changes. Apply spelling patch from Debian maintainer
           Gregor Herman [RT#88745].


0.09     2013-08-22 (SHARYANTO)

         - No functional changes. Reformat Changes to be more conformant to
           CPAN::Changes::Spec (thanks Neil Bowers).


0.08     2012-03-28 (SHARYANTO)

         - peek(), look_back(), has_next(), has_previous() now accept optional
           integer argument for arbitrary lookup, e.g. peek(2) looks at the next
           next item, peek(1) is the same as peek() (implemented by Alexey
           Surikov, github#2).


0.07     2011-09-09 (SHARYANTO)

         - Take over maintenance from Stevan Little.

         - Now uses Dist::Zilla and git.

         - Add lowercase method name aliases (has_next() as well as hasNext(),
           etc). The lowercase method names are now the documented ones.

         - Add iterated() to check whether an iteration has been done (i.e.
           next(), or get_next(), or previous(), etc has been called).


0.06     2005-07-08 (STEVAN)

         - Fixed bug in Array::Iterator::peek(). Thanks to Hugo Cornelis for
           pointing it out

         - added tests for this

         - Added patch from Phillip Moore to support *single element iteration*
           using the hash-ref constructor option. - added tests and docs for
           this (also from Phillip :)


0.05     2004-07-15 (STEVAN)

         - added a getLegnth method and tested it

         - changed how currentIndex deals with index of 0, it now does it
           correctly.

         - made current use currentIndex to get the current index

         - made Array::Iterator more subclass friendly by adding some
           'protected' methods to access some fields with

         - added some subclasses: Array::Iterator::BiDirectional,
           Array::Iterator::Circular, Array::Iterator::Reusable

         - created tests for all these new modules


0.04     2004-05-06 (STEVAN)

         - Changed current and currentIndex to refer to the same value (and
           index) of the last item dispensed by the next method. This is more in
           line with what they should do. Prior to this version they returned
           the current index which was actually the one past the last call to
           next.

         - tested these changes and altered tests which used the old versions.

         - updated documentation to reflect change


0.03     2004-05-02 (STEVAN)

         - Added currentIndex method, and added tests for it.

         - Added getNext method and added tests for it.

         - altered the behavior of peek to not throw an exception.

         - updated all documentation.


0.02     2004-04-12 (STEVAN)

         - error in the Makefile.PL file, no changes on this release


0.01     2004-03-17 (STEVAN)

         - original version; created by h2xs 1.22 with options -X -n
           Array::Iterator


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "A simple class for iterating over Perl arrays",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Array-Iterator",
  "dist_version" => "0.09",
  "license" => "perl",
  "module_name" => "Array::Iterator",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {},
  "script_files" => [],
  "test_requires" => {
    "Test::Exception" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A simple class for iterating over Perl arrays",
  "AUTHOR" => "Steven Haryanto <stevenharyanto\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Array-Iterator",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Array::Iterator",
  "PREREQ_PM" => {},
  "TEST_REQUIRES" => {
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Exception" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.10",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Array/Iterator.pm
package Array::Iterator;

use strict;
use warnings;

our $VERSION = '0.09'; # VERSION

### constructor

sub new {
	my ($_class, @array) = @_;
	(@array)
        || die "Insufficient Arguments : you must provide something to iterate over";
	my $class = ref($_class) || $_class;
	my $_array;
	if (scalar @array == 1) {
		if (ref $array[0] eq "ARRAY") {
		    $_array = $array[0];
		}
		elsif (ref $array[0] eq "HASH") {
		    die "Incorrect type : HASH reference must contain the key __array__"
		        unless exists $array[0]->{__array__};
		    die "Incorrect type : __array__ value must be an ARRAY reference"
		        unless ref $array[0]->{__array__} eq 'ARRAY';
		    $_array = $array[0]->{__array__};
		}
		else {
		    die "Incorrect Type : the argument must be an array or hash reference";
		}
	}
	else {
		$_array = \@array;
	}
	my $iterator = {
        _current_index => 0,
        _length => 0,
        _iteratee => [],
        _iterated => 0,
        };
	bless($iterator, $class);
	$iterator->_init(scalar(@{$_array}), $_array);
	return $iterator;
}

### methods

# private methods

sub _init {
	my ($self, $length, $iteratee) = @_;
	(defined($length) && defined($iteratee))
        || die "Insufficient Arguments : you must provide an length and an iteratee";
	$self->{_current_index} = 0;
	$self->{_length} = $length;
	$self->{_iteratee} = $iteratee;
}

# protected method

# this can be used in a subclass to access the value

# we need to alter this so its an lvalue
sub _current_index : lvalue {
    (UNIVERSAL::isa((caller)[0], __PACKAGE__))
        || die "Illegal Operation : This method can only be called by a subclass";
    $_[0]->{_current_index}
}

# this we should never need to alter
# so we dont make it a lvalue
sub _iteratee {
    (UNIVERSAL::isa((caller)[0], __PACKAGE__))
        || die "Illegal Operation : This method can only be called by a subclass";
    $_[0]->{_iteratee}
}

# we move this from a private method
# to a protected one, and check our access
# as well
sub _getItem {
    (UNIVERSAL::isa((caller)[0], __PACKAGE__))
        || die "Illegal Operation : This method can only be called by a subclass";
	my ($self, $iteratee, $index) = @_;
	return $iteratee->[$index];
}

sub _get_item { my $self = shift; $self->_getItem(@_) }

# we need to alter this so its an lvalue
sub _iterated : lvalue {
    (UNIVERSAL::isa((caller)[0], __PACKAGE__))
        || die "Illegal Operation : This method can only be called by a subclass";
    $_[0]->{_iterated}
}

# public methods

# this defines the interface
# an iterator object will have

sub iterated {
    my ($self) = @_;
    return $self->{_iterated};
}

sub has_next {
	my ($self, $n) = @_;

    if(not defined $n) { $n = 1 }
    elsif(not $n)      { die "has_next(0) doesn't make sense, did you mean current()?" }
    elsif($n < 0)      { die "has_next() with negative argument doesn't make sense, perhaps you should use a BiDirectional iterator" }

    my $idx = $self->{_current_index} + ($n - 1);

	return ($idx < $self->{_length}) ? 1 : 0;
}

sub hasNext { my $self = shift; $self->has_next(@_) }

sub next {
	my ($self) = @_;
    ($self->{_current_index} < $self->{_length})
        || die "Out Of Bounds : no more elements";
        $self->{_iterated} = 1;
	return $self->_getItem($self->{_iteratee}, $self->{_current_index}++);
}

sub get_next {
	my ($self) = @_;
        $self->{_iterated} = 1;
    return undef unless ($self->{_current_index} < $self->{_length});
	return $self->_getItem($self->{_iteratee}, $self->{_current_index}++);
}

sub getNext { my $self = shift; $self->get_next(@_) }

sub peek {
	my ($self, $n) = @_;

    if(not defined $n) { $n = 1 }
    elsif(not $n)      { die "peek(0) doesn't make sense, did you mean get_next()?" }
    elsif($n < 0)      { die "peek() with negative argument doesn't make sense, perhaps you should use a BiDirectional iterator" }

    my $idx = $self->{_current_index} + ($n - 1);

    return undef unless ($idx < $self->{_length});
	return $self->_getItem($self->{_iteratee}, $idx);
}

sub current {
	my ($self) = @_;
	return $self->_getItem($self->{_iteratee}, $self->currentIndex());
}

sub current_index {
	my ($self) = @_;
	return ($self->{_current_index} != 0) ? $self->{_current_index} - 1 : 0;
}

sub currentIndex { my $self = shift; $self->current_index(@_) }

sub get_length {
    my ($self) = @_;
    return $self->{_length};
}

sub getLength { my $self = shift; $self->get_length(@_) }

1;
#ABSTRACT: A simple class for iterating over Perl arrays

__END__

=pod

=head1 NAME

Array::Iterator - A simple class for iterating over Perl arrays

=head1 VERSION

version 0.09

=head1 SYNOPSIS

  use Array::Iterator;

  # create an iterator with an array
  my $i = Array::Iterator->new(1 .. 100);

  # create an iterator with an array reference
  my $i = Array::Iterator->new(\@array);

  # create an iterator with a hash reference
  my $i = Array::Iterator->new({ __array__ => \@array });

  # a base iterator example
  while ($i->has_next()) {
      if ($i->peek() < 50) {
          # ... do something because
          # the next element is over 50
      }
      my $current = $i->next();
      # ... do something with current
  }

  # shortcut style
  my @accumulation;
  push @accumulation => { item => $iterator->next() } while $iterator->has_next();

  # C++ ish style iterator
  for (my $i = Array::Iterator->new(@array); $i->has_next(); $i->next()) {
    my $current = $i->current();
    # .. do something with current
  }

  # common perl iterator idiom
  my $current;
  while ($current = $i->get_next()) {
    # ... do something with $current
  }

=head1 DESCRIPTION

This class provides a very simple iterator interface. It is is uni-directional
and can only be used once. It provides no means of reverseing or reseting the
iterator. It is not recommended to alter the array during iteration, however
no attempt is made to enforce this (although I will if I can find an efficient
means of doing so). This class only intends to provide a clear and simple
means of generic iteration, nothing more (yet).

=for Pod::Coverage .+

=head1 METHODS

=head2 Public Methods

=over 4

=item B<new (@array | $array_ref | $hash_ref)>

The constructor can be passed either a plain perl array, an array reference,
or a hash reference (with the array specified as a single key off the hash,
__array__). Single element arrays are not supported by either of the first
two calling conventions, since it is not possible to distinguish between an
array of a single element which happens to be an array reference, and an
array reference of a single element, thus previous versions of the constructor
would raise an exception. If you expect to pass arrays to the constructor which
may have only a single element, then the array can be passed as the element
of a HASH reference, with the key, __array__:

  my $i = Array::Iterator->new({ __array__ => \@array });

=item B<has_next([$n])>

This methods returns a boolean. True (1) if there are still more elements in
the iterator, false (0) if there are not.

Takes an optional positive integer (E<gt> 0) that specifies the position you
want to check. This allows you to check if there an element at arbitrary position.
Think of it as an ordinal number you want to check:

  $i->has_next(2);  # 2nd next element
  $i->has_next(10); # 10th next element

Note that C<has_next(1)> is the same as C<has_next()>.

Throws an exception if C<$n> E<lt>= 0.

=item B<next>

This method returns the next item in the iterator, be sure to only call this
once per iteration as it will advance the index pointer to the next item. If
this method is called after all elements have been exhausted, an exception
will be thrown.

=item B<get_next>

This method returns the next item in the iterator, be sure to only call this
once per iteration as it will advance the index pointer to the next item. If
this method is called after all elements have been exhausted, it will return
undef.

This method was added to allow for a faily common perl iterator idiom of:

  my $current;
  while ($current = $i->get_next()) {
      ...
  }

In this the loop terminates once C<$current> is assigned to a false value.
The only problem with this idiom for me is that it does not allow for
undefined or false values in the iterator. Of course, if this fits your
data, then there is no problem. Otherwise I would recommend the C<has_next>/C<next>
idiom instead.

=item B<peek([$n])>

This method can be used to peek ahead at the next item in the iterator. It
is non-destructuve, meaning it does not advance the internal pointer. If
this method is called and attempts to reach beyond the bounds of the iterator,
it will return undef.

Takes an optional positive integer (E<gt> 0) that specifies how far ahead you want to peek:

  $i->peek(2);  # gives you 2nd next element
  $i->peek(10); # gives you 10th next element

Note that C<peek(1)> is the same as C<peek()>.

Throws an exception if C<$n> E<lt>= 0.

B<NOTE:> Prior to version 0.03 this method would throw an exception if called
out of bounds. I decided this was not a good practice, as it made it difficult
to be able to peek ahead effectively. This not the case when calling with an argument
that is E<lt>= 0 though, as it's clearly a sign of incorrect usage.

=item B<current>

This method can be used to get the current item in the iterator. It is non-destructive,
meaning that it does not advance the internal pointer. This value will match the
last value dispensed by C<next> or C<get_next>.

=item B<current_index>

This method can be used to get the current index in the iterator. It is non-destructive,
meaning that it does not advance the internal pointer. This value will match the index
of the last value dispensed by C<next> or C<get_next>.

=item B<get_length>

This is a basic accessor for getting the length of the array being iterated over.

=back

=head2 Protected Methods

These methods are I<protected>, in the Java/C++ sense of the word. They can only be
called internally by subclasses of Array::Iterator, an exception is thrown if that
condition is violated. They are documented here only for people interested in
subclassing Array::Iterator.

=over 4

=item B<_current_index>

An lvalue-ed subroutine which allows access to the iterator's internal pointer.

=item B<_iteratee>

This returns the item being iteratated over, in our case an array.

=item B<_get_item ($iteratee, $index)>

This method is used by all other routines to access items with. Given the iteratee
and an index, it will return the item being stored in the C<$iteratee> at the index
of C<$index>.

=back

=head1 TO DO

=over 4

=item Improve BiDirectional Test suite

I want to test the back and forth a little more, make sure they work well with
one another.

=item Other Iterators

Array::Iterator::BiDirectional::Circular, Array::Iterator::Skipable and
Array::Iterator::BiDirectional::Skipable are just a few ideas I have had. I am going
to hold off for now until I am sure they are actually useful.

=back

=head1 BUGS

None that I am aware of. The code is pretty thoroughly tested (see L<CODE COVERAGE> below)
and is based on an (non-publicly released) module which I had used in production systems
for about 2 years without incident. Of course, if you find a bug, let me know, and I will
be sure to fix it.

=head1 CODE COVERAGE

I use B<Devel::Cover> to test the code coverage of my tests, below is the B<Devel::Cover>
report on this module's test suite.

 ------------------------------- ------ ------ ------ ------ ------ ------ ------
 File                              stmt   bran   cond    sub    pod   time  total
 ------------------------------- ------ ------ ------ ------ ------ ------ ------
 Array/Iterator.pm                100.0  100.0   66.7  100.0  100.0   67.6   98.2
 Array/Iterator/BiDirectional.pm  100.0  100.0    n/a  100.0  100.0   20.2  100.0
 Array/Iterator/Circular.pm       100.0  100.0    n/a  100.0  100.0    7.1  100.0
 Array/Iterator/Reusable.pm       100.0    n/a    n/a  100.0  100.0    5.0  100.0
 ------------------------------- ------ ------ ------ ------ ------ ------ ------
 Total                            100.0  100.0   66.7  100.0  100.0  100.0   99.0
 ------------------------------- ------ ------ ------ ------ ------ ------ ------

=head1 SEE ALSO

This module now includes several subclasses of Array::Iterator which add certain behaviors
to Array::Iterator, they are:

=over 4

=item B<Array::Iterator::BiDirectional>

Adds the ability to move backwards and forwards through the array.

=item B<Array::Iterator::Circular>

When this iterator reaches the end of its list, it will loop back to the start again.

=item B<Array::Iterator::Reusable>

This iterator can be reset to its begining and used again.

=back

The Design Patterns book by the Gang of Four, specifically the Iterator pattern.

Some of the interface for this class is based upon the Java Iterator interface.

=head1 OTHER ITERATOR MODULES

There are a number of modules on CPAN with the word Iterator in them. Most of them are
actually iterators included inside other modules, and only really useful within that
parent modules context. There are however some other modules out there that are just
for pure iteration. I have provided a list below of the ones I have found, if perhaps
you don't happen to like the way I do it.

=over 4

=item B<Tie::Array::Iterable>

This module ties the array, something we do not do. But it also makes an attempt to
account for, and allow the array to be changed during iteration. It accomplishes this
control because the underlying array is tied. As we all know, tie-ing things can be a
performance issue, but if you need what this module provides, then it will likely be
an acceptable compromise. Array::Iterator makes no attempt to deal with this mid-iteration
manipulation problem. In fact it is recommened to not alter your array with Array::Iterator,
and if possible we will enforce this in later versions.

=item B<Data::Iter>

This module allows for simple iteratation over both hashes and arrays. It does it by
importing several functions which can be used to loop over either type (hash or array)
in the same way. It is an interesting module, it differs from Array::Iterator in
paradigm (Array::Iterator is more OO) as well as in intent.

=item B<Class::Iterator>

This is essentially a wrapper around a closure based iterator. This method can be very
flexible, but at times is difficult to manage due to the inherent complextity of using
closures. I actually was a closure-as-iterator fan for a while, but eventually moved
away from it in favor of the more plain vanilla means of iteration, like that found
Array::Iterator.

=item B<Class::Iter>

This is part of the Class::Visitor module, and is a Visitor and Iterator extensions to
Class::Template. Array::Iterator is a standalone module not associated with others.

=item B<Data::Iterator::EasyObj>

Data::Iterator::EasyObj makes your array of arrays into iterator objects. It also has
the ability to further nest additional data structures including Data::Iterator::EasyObj
objects. Array::Iterator is one dimensional only, and does not attempt to do many of
the more advanced features of this module.

=back

=head1 ACKNOWLEDGEMENTS

=over 4

=item Thanks to Hugo Cornelis for pointing out a bug in C<peek()>

=item Thanks to Phillip Moore for providing the patch to allow single element iteration
through the hash-ref constructor parameter.

=back

=head1 ORIGINAL AUTHOR

stevan little, E<lt>stevan@iinteractive.comE<gt>

=head1 ORIGINAL COPYRIGHT AND LICENSE

Copyright 2004, 2005 by Infinity Interactive, Inc.

L<http://www.iinteractive.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
